test/test_game.py

```diff
@@ -0,0 +1,229 @@
+import unittest
+
+from rock_paper_scissors.core.exc import InvalidRoundsError, InvalidChoiceError
+from rock_paper_scissors.core.game import Game
+from rock_paper_scissors.core.moves import Paper, Rock
+from rock_paper_scissors.core.user import User
+from rock_paper_scissors.core.valid_choices import VALID_CHOICES
+
+
+class GameInitTestCase(unittest.TestCase):
+    def test_player1_and_player2_and_starting_rounds_types(self):
+        player1 = 'Joe'
+        player2 = 'Jane'
+        game = Game(player1, player2)
+
+        self.assertIsInstance(game.player1, str)
+        self.assertIsInstance(game.player2, str)
+        self.assertEqual(game._rounds, 0)
+
+
+class GameSetRoundsTestCase(unittest.TestCase):
+    def setUp(self):
+        self.game = Game('player1', 'player2')
+
+    def test_valid_rounds_input(self):
+        value = self.game.rounds = 11
+
+        self.assertEqual(value, 11)
+
+    def test_raises_error_if_round_not_an_int(self):
+        with self.assertRaises(InvalidRoundsError) as ctx:
+            self.game.rounds = 'a'
+
+        self.assertEqual(
+            str(ctx.exception),
+            'Choice must be an integer'
+        )
+
+    def test_raises_error_if_value_is_not_odd(self):
+        with self.assertRaises(InvalidRoundsError) as ctx:
+            self.game.rounds = 2
+
+        self.assertEqual(
+            str(ctx.exception),
+            'Choice must be an odd number'
+        )
+
+    def test_raises_error_if_value_is_not_positive(self):
+        with self.assertRaises(InvalidRoundsError) as ctx:
+            self.game.rounds = -1
+
+        self.assertEqual(
+            str(ctx.exception),
+            'Choice must be greater than zero'
+        )
+
+
+class GameGetRoundsTestCase(unittest.TestCase):
+    def test_return_rounds_of_game(self):
+        self.game = Game('player1', 'player2')
+        self.game._rounds = 1
+
+        self.assertEqual(self.game.rounds, 1)
+
+
+class GameValidChoicesTestCase(unittest.TestCase):
+    def test_assert_tuple_of_valid_choices(self):
+        self.game = Game('player1', 'player2')
+        game_valid_choices = self.game.valid_choices
+
+        self.assertIsInstance(game_valid_choices, tuple)
+        self.assertCountEqual(game_valid_choices, tuple(VALID_CHOICES.keys()))
+
+
+class GameConvertPlayerInputToMove(unittest.TestCase):
+    def test_user_input_is_a_valid_string_for_conversion_to_move(self):
+        self.game = Game('player1', 'player2')
+        expected = VALID_CHOICES['rock']
+        result = self.game.convert_input_to_move('rock')
+
+        self.assertEqual(expected, result)
+
+    def test_user_input_is_not_a_valid_string_for_conversion_to_object(self):
+        self.assertRaises(InvalidChoiceError, Game.convert_input_to_move, 'fish')
+
+
+class GamePlayLogicTestCase(unittest.TestCase):
```

Comment:
@TylerHendrickson for this test case, am I being too granular with the unit tests? Especially with no point allocation such as `def test_player1_gets_no_point_if_player2_move_greater_than_player1_move(self)`

Reply:
I think that as long as you have one test dedicated to each logical branch of execution, it's not too granular. In cases where you have more than one test dedicated to the same logical branch, it makes sense to combine them, such as in the case of `test_player1_gets_no_point_if_player2_move_greater_than_player1_move` and `test_player2_does_not_get_point_if_player1_move_equals_player2_move`.

As a general rule, if two tests are in fact testing separate logical branches but the two tests are completely coupled (eg. one cannot pass unless the other also passes, and similarly cannot fail unless the other test also fails), then it _may_ indicate that the implementation itself is overly complex and could be simplified.